### PR TITLE
Graceful sidecar support

### PR DESCRIPTION
**Pull request by joseblas**
# Changes

TL;DR Tekton Pipelines now gracefully start and stop with injected sidecars running alongside TaskRun pods.

Long Version :-

TaskRun pods can have sidecars injected by Admission Controllers. This is common practice with Istio where a proxy sidecar is injected into pods so that they can be included as part of a service mesh. Unfortunately there is no built-in Kubernetes mechanism to set the lifetime of a sidecar container to match that of the pod's "primary" container. In practice this means that pods injected with sidecars will not stop until both the primary and all sidecar containers are stopped.

Prior to this commit any non-istio injected sidecars would cause a TaskRun pod to run forever, even when all Step containers of a Task had completed.  Istio injection was also disabled completely on TR pods. This commit introduces mechanisms to start and stop sidecars gracefully meaning that a) Step containers wait until all sidecars are in a ready state before starting and b) Sidecar containers are stopped when a TaskRun's Step containers are done.  Additionally, the annotation that prevented Istio sidecar injection has been removed.  Istio now adds sidecars to TaskRun pods as expected.

No end-to-end tests have been added as part of this PR because doing so would require Istio to be enabled on the CI cluster (or a dummy Admission Controller / Mutating Webhook thing to be written just for injecting containers into test Pods). The Istio requirement would put an undue burden on those users running tests in non-GKE environments. It's currently planned for a follow-up PR to introduce a) user-defined sidecars and b) a Tekton-injected sidecar that performs logging. Once that work is complete it will be much simpler to e2e test this sidecar code.

For reference, see the POC PR that was originally created here: https://github.com/tektoncd/pipeline/pull/728

For the design proposal and associated discussion around the original POC please see https://github.com/tektoncd/pipeline/issues/727

For the remaining work, test coverage, code organization, etc, see https://github.com/tektoncd/pipeline/issues/926

# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

# Release Notes

```
Sidecar containers injected by, for example, Istio, will now gracefully start up and shutdown in tandem with a Task Steps' containers.
```

### CODE REVIEW

**Comment by sbwsg** [comment]
Hm, from the integration tests:

```
I0531 14:26:40.473] 2019/05/31 14:26:40 main.go:761: --gcp-project is missing, trying to fetch a project from boskos.
I0531 14:26:40.474] (for local runs please set --gcp-project to your dev project)
I0531 14:26:40.474] 2019/05/31 14:26:40 main.go:773: provider gke, will acquire project type gke-project from boskos
I0531 14:26:40.484] 2019/05/31 14:26:40 main.go:312: Something went wrong: failed to prepare test environment: --provider=gke boskos failed to acquire project: resource not found
```

This looks suspiciously like a configuration or platform issue.

**Comment by sbwsg** [comment]
Yeah I'm in favor of that if you can make it work, but defer to @bobcatfish and also @vdemeester in case they have concerns given our existing test infra.  How would this work if someone is trying to run e2e tests locally or on a non-GKE Kubernetes provider?

**Comment by joseblas** [comment]
Istio/sdecar related  tests could be skipped if there is no Istio namespace.

**Comment by vdemeester** [comment]
Having to install istio just for the e2e tests seems a bit much (and heavy) I feel. Discussed a bit with @bobcatfish and @sbwsg and we fell we can defer the e2e a bit (until we work on *logging sidecar* and thus using this as e2e tests for sidecars).

**Comment by sbwsg** [comment]
Hey @dicarlo2 I think we're pretty close to having your sidecar PR mergeable here. If you've got some time would you mind reviewing our updates to your work and verifying that the documentation and code comments we've added are accurate and gel with your intentions?  Cheers!

**Comment by dicarlo2** [comment]
LGTM, sorry for the delayed response!

**Comment by sbwsg** [comment]
Thanks @dicarlo2 !  I've modified the cla label on this PR to indicate that everything is good.

File: docs/taskruns.md [hunk 1]
```diff
@@ -536,6 +537,29 @@ of the `Task` resource object.
...
+to inject an envoy proxy in to pods so that they can be included as part of
+Istio's service mesh.
+
+Tekton will happily work with sidecars injected into a TaskRun's
+pods but the behaviour is a bit nuanced: When a TaskRun's steps are complete
+any sidecar containers running inside the Pod will be terminated. In
+order to terminate the sidecars they will be restarted with a new
+"nop" image that quickly exits. The result will be that your TaskRun's
+Pod will include the sidecar container with a Retry Count of 1 and
+with a different container image than you might be expecting.
```
Comment (bobcatfish) [inline docs/taskruns.md]: :+1: nice docs! :D

File: cmd/entrypoint/main.go [hunk 2]
```diff
@@ -75,14 +77,16 @@ type RealWaiter struct{}
 
 var _ entrypoint.Waiter = (*RealWaiter)(nil)
 
-func (*RealWaiter) Wait(file string) error {
+func (*RealWaiter) Wait(file string, expectContent bool) error {
 	if file == "" {
 		return nil
 	}
 	for ; ; time.Sleep(time.Second) {
 		// Watch for the post file
```
Comment (bobcatfish) [inline cmd/entrypoint/main.go]: can we expand this comment here to elaborate on what we expect to populate the file, and why, and with what?

File: cmd/entrypoint/main.go [hunk 3]
```diff
@@ -75,14 +77,16 @@ type RealWaiter struct{}
...
-		if _, err := os.Stat(file); err == nil {
-			return nil
+		if info, err := os.Stat(file); err == nil {
+			if !expectContent || info.Size() > 0 {
+				return nil
+			}
 		} else if !os.IsNotExist(err) {
 			return xerrors.Errorf("Waiting for %q: %w", file, err)
 		}
```
Comment (bobcatfish) [inline cmd/entrypoint/main.go]: I'd love to see some/any unit tests for this if possible :innocent: 

(i think what i'd do is take the bulk of the `Wait` function, put it in a separate function - probably in  pkg/entrypoint/entrypointer.go - and test that)

File: pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go [hunk 4]
```diff
@@ -135,34 +142,45 @@ func RedirectStep(cache *Cache, stepNum int, step *corev1.Container, kubeclient
 	step.Args = GetArgs(stepNum, step.Command, step.Args)
 	step.Command = []string{BinaryLocation}
 	step.VolumeMounts = append(step.VolumeMounts, toolsMount)
+	if stepNum == 0 {
+		step.VolumeMounts = append(step.VolumeMounts, downwardMount)
+	}
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go]: can we have a comment here explaining why this is happening, why it's only for step 0, what it's for, etc.?

File: pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go [hunk 5]
```diff
@@ -135,34 +142,45 @@ func RedirectStep(cache *Cache, stepNum int, step *corev1.Container, kubeclient
...
 }
 
 // GetArgs returns the arguments that should be specified for the step which has been wrapped
 // such that it will execute our custom entrypoint instead of the user provided Command and Args.
 func GetArgs(stepNum int, commands, args []string) []string {
-	waitFile := fmt.Sprintf("%s/%s", MountPoint, strconv.Itoa(stepNum-1))
-	if stepNum == 0 {
-		waitFile = ""
-	}
+	waitFile := getWaitFile(stepNum)
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go]: woot :D

File: pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go [hunk 6]
```diff
@@ -135,34 +142,45 @@ func RedirectStep(cache *Cache, stepNum int, step *corev1.Container, kubeclient
...
+	}
+	if stepNum == 0 {
+		argsForEntrypoint = append(argsForEntrypoint, "-wait_file_content")
+	}
+	argsForEntrypoint = append(argsForEntrypoint, "-entrypoint")
+	argsForEntrypoint = append(argsForEntrypoint, commands...)
 	// TODO: what if Command has multiple elements, do we need "--" between command and args?
 	argsForEntrypoint = append(argsForEntrypoint, "--")
 	return append(argsForEntrypoint, args...)
 }
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/entrypoint/entrypoint.go]: pretty cool that you were able to clean this up - and the tests didnt even have to change :) nice improvements!

File: pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go [hunk 7]
```diff
@@ -435,3 +420,42 @@ func TestMakeWorkingDirScript(t *testing.T) {
...
+		pod:                        &corev1.Pod{},
+		updateFunc:                 func(p *corev1.Pod) (*corev1.Pod, error) { return p, nil },
+		expectedAnnotationsContain: map[string]string{ReadyAnnotation: readyAnnotationValue},
+		expectedErr:                nil,
+	}, {
+		desc:                       "errors experienced during update are returned",
+		pod:                        &corev1.Pod{},
+		updateFunc:                 func(p *corev1.Pod) (*corev1.Pod, error) { return p, testerror },
+		expectedAnnotationsContain: map[string]string{ReadyAnnotation: readyAnnotationValue},
+		expectedErr:                testerror,
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: instead of using `expectedErr` i find it much clearer to have two completely different tests, one for success conditions and one for error conditions (less potential for something to go wrong and also easier to read)

File: pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go [hunk 8]
```diff
@@ -435,3 +420,42 @@ func TestMakeWorkingDirScript(t *testing.T) {
...
+	}} {
+		t.Run(c.desc, func(t *testing.T) {
+			err := AddReadyAnnotation(c.pod, c.updateFunc)
+			if err != c.expectedErr {
+				t.Errorf("expected error %v but received %v", c.expectedErr, err)
+			}
+			for a, v := range c.expectedAnnotationsContain {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: maybe we can avoid looping here since the only test cases we have only have one annotation?

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 9]
```diff
@@ -61,6 +62,10 @@ const (
 	// is just starting to be reconciled
 	reasonRunning = "Running"
 
+	// reasonBuilding indicates that the reasono for the in progress status is that the TaskRun
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: typo `reasono`

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 10]
```diff
@@ -61,6 +62,10 @@ const (
...
+	// is just being built
+	reasonBuilding = "Building"
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: what does "buliding" mean in this context? since this whole project started from "knative build", the term Build is pretty overloaded - and i dont think we're actually building anything here :thinking:  is there a more specific term we could use?

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 11]
```diff
@@ -82,6 +87,10 @@ const (
 	imageDigestExporterContainerName = "step-image-digest-exporter"
 )
 
+var (
+	nopImage = flag.String("nop-image", "override-with-nop:latest", "The container image used to kill sidecars")
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: can we update the nopImage `cmd` source to explain why we have this around still? (since we don't need to include it as an actual step now that we're not using init containers for steps)

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 12]
```diff
@@ -329,18 +342,26 @@ func (c *Reconciler) reconcile(ctx context.Context, tr *v1alpha1.TaskRun) error
...
 
 	c.Logger.Infof("Successfully reconciled taskrun %s/%s with status: %#v", tr.Name, tr.Namespace, after)
 
 	return nil
 }
 
-func updateStatusFromPod(taskRun *v1alpha1.TaskRun, pod *corev1.Pod, resourceLister listers.PipelineResourceLister, kubeclient kubernetes.Interface, logger *zap.SugaredLogger) {
+// updateStatusFromPod modifies the task run status based on the pod and then returns true if the pod is running and
+// all sidecars are ready
+func updateStatusFromPod(taskRun *v1alpha1.TaskRun, pod *corev1.Pod, resourceLister listers.PipelineResourceLister, kubeclient kubernetes.Interface, logger *zap.SugaredLogger) bool {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: especially now that this function is changing substantially, can it be moved out into a different package and have some unit tests added? (that might be a lot to do all at once but we could start with just a couple simple test - it would be more than we have currently!)

it would probably also mean you'd have to break this function up a bit - which I think would be good to do cuz it's getting a bit complex

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 13]
```diff
@@ -354,54 +375,93 @@ func updateStatusFromPod(taskRun *v1alpha1.TaskRun, pod *corev1.Pod, resourceLis
 	taskRun.Status.PodName = pod.Name
 
 	taskRun.Status.Steps = []v1alpha1.StepState{}
+	// Might be complete if we're only running because of a sidecar
+	maybeComplete := len(pod.Status.ContainerStatuses) > 0 && pod.Status.Phase == corev1.PodRunning
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: hm im a bit confused about what "maybe complete" means here - would calling this something like `stepsComplete` make sense?

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 14]
```diff
@@ -664,6 +758,45 @@ func getExceededResourcesMessage(tr *v1alpha1.TaskRun) string {
...
+// killSidecars stops all sidecar containers running inside a taskrun's pod.
+// A container is determined to be a sidecar if it is currently running - this
+// func is only expected to be called after a TaskRun completes and so all Task
+// containers should have stopped already.
+//
+// A sidecar is killed by replacing its current container image with the nop
+// image, which in turn quickly exits. If the sidedcar defines a command then
+// it will exit with a non-zero status. When we check for TaskRun success we
+// have to check for the containers we care about - not the final Pod status.
+func (c *Reconciler) killSidecars(tr *v1alpha1.TaskRun) error {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: can this also be moved out into a different package? (it means you'll have to pass in a function for getting pods, which can easily be mocked and makes test setup a bit easier and clearer)

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go [hunk 15]
```diff
@@ -229,6 +228,14 @@ func TryGetPod(taskRunStatus v1alpha1.TaskRunStatus, gp GetPod) (*corev1.Pod, er
 // MakePod converts TaskRun and TaskSpec objects to a Pod which implements the taskrun specified
 // by the supplied CRD.
 func MakePod(taskRun *v1alpha1.TaskRun, taskSpec v1alpha1.TaskSpec, kubeclient kubernetes.Interface, cache *entrypoint.Cache, logger *zap.SugaredLogger) (*corev1.Pod, error) {
+	// Copy annotations on the build through to the underlying pod to allow users
+	// to specify pod annotations.
+	annotations := map[string]string{}
+	for key, val := range taskRun.Annotations {
+		annotations[key] = val
+	}
+	annotations[ReadyAnnotation] = ""
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: can this code be a separate function with unit tests?

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go [hunk 16]
```diff
@@ -664,6 +758,45 @@ func getExceededResourcesMessage(tr *v1alpha1.TaskRun) string {
...
+			return nil
+		}
+
+		return err
+	}
+
+	updated := false
+	if pod.Status.Phase == corev1.PodRunning {
+		for i, s := range pod.Status.ContainerStatuses {
+			if s.State.Running != nil && pod.Spec.Containers[i].Image != *nopImage {
```
Comment (sbwsg) [inline pkg/reconciler/v1alpha1/taskrun/taskrun.go]: Bug here: the order of the pod's `ContainerStatuses` field doesn't always match the order of the pod's `Containers` but we're expecting that order to be consistent in order to stop the right container.

**Comment by sbwsg** [comment]
> would you mind adding an issue to the backlog about adding support to TaskRuns to add sidecars directly?

I've added an issue to capture the user-specified sidecars feature: https://github.com/tektoncd/pipeline/issues/980

File: cmd/entrypoint/main.go (hunk 2 above)
Comment (sbwsg) [reply cmd/entrypoint/main.go]: I've removed the two inline comments and added a godoc to this func that explains the behaviour and edge cases.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 10 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: hrm. We were already using the string "building" as the `reason` for a TaskRun with a running Pod.  This change simply put the string into a variable so that it could be referred to from unit tests without using the literal string in multiple places.

I'd be happy to change this but it doesn't feel related to this PR.  Modifying the TaskRun's Status could also be considered an API change? I'd want to make sure it's isolated in its own commit at least so that we can revert it easily.  I'll make a separate PR for this if we can come up with a better descriptor.  "Executing" might work instead?  "Running" is already taken for a TaskRun that has started up but doesn't yet have a `ConditionSucceeded` status.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 12 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: I've moved it out along with its associated unit tests into a new `taskrun/status` package.  I've also moved all of the `reason...` variables into constants inside this package as well.  A couple of other functions needed to move along with it.  The diff's quite a bit bigger now but it felt good to refactor it.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 13 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: I've refactored this part of the function a little bit to make it slightly more readable.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 12 above)
Comment (bobcatfish) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: awesome!!

quick heads up that it looks like the `status` package will be added in https://github.com/tektoncd/pipeline/pull/963 as well (more evidence that it's a good name imo!) so there might be some conflicts for someone

File: docs/developers/README.md [hunk 17]
```diff
@@ -155,3 +156,24 @@ such as the following:
 - The environment variable HOME is set to `/builder/home`, used by the builder
   tools and injected on into all of the step containers
 - Default location for output-images `/builder/output-images`
+
+## Handling of injected sidecars
+
+The [Downward API](TODO: link) is used to project an annotation on the
```
Comment (bobcatfish) [inline docs/developers/README.md]: https://kubernetes.io/docs/tasks/inject-data-application/downward-api-volume-expose-pod-information/#the-downward-api seems okay but I tried to find docs that actually really explain how this works and i couldn't find anything very good :(

File: docs/developers/README.md [hunk 18]
```diff
@@ -155,3 +156,24 @@ such as the following:
...
+annotation is populated with string "READY", which in turn populates the
+Downward API projected file. The entrypoint binary recognizes
+that the projected file has a non-zero size and allows the Task's steps to
+begin.
+
+On completion of all steps in a Task the TaskRun reconciler stops any
+sidecar containers. The `Image` field of any sidecar containers is swapped
+to the nop image. Kubernetes observes the change and relaunches the container
+with updated container image. The nop container image exits. The container
+is considered
```
Comment (bobcatfish) [inline docs/developers/README.md]: awesome!! thanks for writing this up, very clear! and a pretty nice design imo :D (*high five* @dicarlo2 :D)

File: pkg/reconciler/v1alpha1/taskrun/sidecars/stop.go [hunk 19]
```diff
@@ -0,0 +1,46 @@
...
+// Stop stops all sidecar containers inside a pod. A container is considered
+// to be a sidecar if it is currently running. This func is only expected to
+// be called after a TaskRun completes and all Step containers Step containers
+// have already stopped.
+//
+// A sidecar is killed by replacing its current container image with the nop
+// image, which in turn quickly exits. If the sidedcar defines a command then
+// it will exit with a non-zero status. When we check for TaskRun success we
+// have to check for the containers we care about - not the final Pod status.
+func Stop(pod *corev1.Pod, updatePod UpdatePod) error {
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/sidecars/stop.go]: nice package and function name! `sidecars.Stop` :heart_eyes_cat:

File: pkg/reconciler/v1alpha1/taskrun/sidecars/stop_test.go [hunk 20]
```diff
@@ -0,0 +1,102 @@
...
+			if err := Stop(pod, updatePod); err != nil {
+				t.Errorf("error stopping sidecar: %v", err)
+			}
+			sidecarIdx := len(pod.Spec.Containers) - 1
+			if pod.Spec.Containers[sidecarIdx].Image != tc.expectedImage {
+				t.Errorf("expected sidecar image %q, actual: %q", tc.expectedImage, pod.Spec.Containers[sidecarIdx].Image)
+			}
+		})
+	}
+}
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/sidecars/stop_test.go]: :heart_eyes_cat: :heart_eyes_cat: :heart_eyes_cat:

File: pkg/reconciler/v1alpha1/taskrun/status/pod.go [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+					logger.Errorf("Error getting output from image-digest-exporter for %s/%s: %s", taskRun.Name, taskRun.Namespace, err)
+				}
+				err = resources.UpdateTaskRunStatusWithResourceResult(taskRun, logContent)
+				if err != nil {
+					logger.Errorf("Error getting output from image-digest-exporter for %s/%s: %s", taskRun.Name, taskRun.Namespace, err)
+				}
+			}
+		}
+	}
+}
```
Comment (bobcatfish) [inline pkg/reconciler/v1alpha1/taskrun/status/pod.go]: oh wow, significant improvement!!! thanks so much :bowing_woman:

File: docs/developers/README.md (hunk 18 above)
Comment (vdemeester) [inline docs/developers/README.md]: `The container is considered` <- isn't something missing ?

File: docs/taskruns.md [hunk 22]
```diff
@@ -536,6 +537,29 @@ of the `Task` resource object.
...
+Tekton doesn't provide a mechanism to specify sidecars for Task steps
+but it's still possible for sidecars to be added to your Pods:
+[Admission Controllers](https://kubernetes.io/docs/reference/access-authn-authz/admission-controllers/)
+provide cluster admins a mechanism to inject sidecar containers as Pods launch.
+As a concrete example this is one possible method [used by Istio](https://istio.io/docs/setup/kubernetes/additional-setup/sidecar-injection/#automatic-sidecar-injection)
+to inject an envoy proxy in to pods so that they can be included as part of
+Istio's service mesh.
+
+Tekton will happily work with sidecars injected into a TaskRun's
+pods but the behaviour is a bit nuanced: When a TaskRun's steps are complete
```
Comment (vdemeester) [inline docs/taskruns.md]: `when TaskRun's steps …`

File: pkg/reconciler/v1alpha1/taskrun/status/pod.go [hunk 23]
```diff
@@ -0,0 +1,226 @@
...
+				ContainerState: *s.State.DeepCopy(),
+				Name:           resources.TrimContainerNamePrefix(s.Name),
+			})
+		}
+	}
+
+	// Complete if we did not find a step that is not complete, or the pod is in a definitely complete phase
+	complete := areStepsComplete(pod) || pod.Status.Phase == corev1.PodSucceeded || pod.Status.Phase == corev1.PodFailed
+
+	if complete {
```
Comment (vdemeester) [inline pkg/reconciler/v1alpha1/taskrun/status/pod.go]: nit: This could be splitted once more (aka `if complete { function } else { function }`)

File: pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go (hunk 7 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: I've separated this into two tests now, one for the successful case and one for the error case.

File: pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go (hunk 8 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod_test.go]: Done, I've removed the hash map looping and just check the single expected annotation.

File: pkg/reconciler/v1alpha1/taskrun/resources/pod.go (hunk 15 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/resources/pod.go]: Good catch.  it turns out there is already a separate function (see `makeAnnotations()` elsewhere in this file) and these few lines are dead code. I've added some unit tests for makeAnnotations.

File: pkg/reconciler/v1alpha1/taskrun/taskrun.go (hunk 11 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/taskrun.go]: I've added a comment to the nop cmd explaining its reason for still existing.

File: docs/developers/README.md (hunk 18 above)
Comment (sbwsg) [reply docs/developers/README.md]: :sweat_smile: oops.  finished this thought in another commit.

File: pkg/reconciler/v1alpha1/taskrun/status/pod.go (hunk 23 above)
Comment (sbwsg) [reply pkg/reconciler/v1alpha1/taskrun/status/pod.go]: nice.  doing this also made me realise that there was some dead code here; the `else` branch was handling pod phases of PodSucceeded and PodFailed but neither of these phases will ever reach the `else` branch since `complete` will be true in both of these cases.

File: cmd/entrypoint/main.go (hunk 3 above)
Comment (sbwsg) [reply cmd/entrypoint/main.go]: I've added unit tests for the `RealWaiter` type.  I wasn't entirely sure how best to break this out into the `entrypoint` package so I've just added them initially alongside `main.go` in a `main_test.go`. Happy to move them elsewhere but might need a bit more guidance on how best to separate things out.

**Comment by sbwsg** [comment]
@bobcatfish @vdemeester thanks for taking a look so far.  It's a big PR so I appreciate the :eyes:! I believe I've tackled all the remaining issues that were left outstanding from yesterday's and today's reviews. Please take a look when you can.  Cheers!

**Comment by sbwsg** [comment]
/hold

Once reviews are done I want to squash before merging.

File: cmd/entrypoint/main_test.go [hunk 24]
```diff
@@ -0,0 +1,92 @@
...
+		close(doneCh)
+	}()
+	ioutil.WriteFile(tmp.Name(), []byte("😺"), 0700)
+	select {
+	case <-doneCh:
+		// Success
+	case <-time.After(2 * waitPollingInterval):
+		t.Errorf("expected Wait() to have detected a non-zero file size by now")
+	}
+}
```
Comment (bobcatfish) [inline cmd/entrypoint/main_test.go]: beautiful!! :sob: 🤩

File: cmd/nop/main.go [hunk 25]
```diff
@@ -14,6 +14,12 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
+// The nop command is a no-op, it simply prints a message and exits. Nop
+// is used to stop sidecar containers in TaskRun Pods. When a Task's Steps
+// are complete any sidecars running alongside the Step containers need
+// to be terminated. Whatever image the sidecars are running is replaced
+// with nop and the sidecar quickly exits.
```
Comment (bobcatfish) [inline cmd/nop/main.go]: nice!

**Comment by sbwsg** [comment]
/hold cancel

I've squashed all the commits together.

**Comment by bobcatfish** [comment]
Thanks for squashing the commits @sbwsg !!

Interestingly https://github.com/tektoncd/pipeline/pull/963 adds a `status` pkg but in a different location than this PR 🤔  I think we might want to combine them together but for now I'm happy to merge this :D

Thanks for the awesome work @sbwsg @joseblas @dicarlo2 !!

**Comment by vdemeester** [comment]
@sbwsg arf, needs a rebase :pray:

**Comment by sbwsg** [comment]
@bobcatfish @vdemeester I've rebased on master and moved my `taskrun/status` changes into the `status` pkg so we only have one.

**Comment by sbwsg** [comment]
ಠ_ಠ

```
main.go:312: Something went wrong: failed to prepare test environment: --provider=gke boskos failed to acquire project: resource not found
```

**Comment by sbwsg** [comment]
:raised_hands:

**Comment by dicarlo2** [comment]
awesome work!
